src/Microsoft.AspNetCore.Cors/Infrastructure/CorsService.cs

```diff
@@ -82,7 +82,7 @@ public CorsResult EvaluatePolicy(HttpContext context, CorsPolicy policy)
         public virtual void EvaluateRequest(HttpContext context, CorsPolicy policy, CorsResult result)
         {
             var origin = context.Request.Headers[CorsConstants.Origin];
-            if (StringValues.IsNullOrEmpty(origin) || !policy.AllowAnyOrigin && !policy.Origins.Contains(origin))
+            if (StringValues.IsNullOrEmpty(origin) || !policy.AllowAnyOrigin && !policy.RegularOrigins.Any(r => r.IsMatch(origin)))
```

Comment:
Actually why do you even need to support regex? Just do the wildcard matching by hand, it'll be more predictable and efficient that way. Also we can start by allowing less things. Opening it up to regex makes it hard to optimize.